packages/monitor-v2/src/price-publisher/PublishPrices.ts

```diff
@@ -0,0 +1,123 @@
+import "@nomiclabs/hardhat-ethers";
+import { paginatedEventQuery } from "@uma/common";
+import { OracleHubEthers, VotingV2Ethers } from "@uma/contracts-node";
+import { ArbitrumParentMessenger } from "@uma/contracts-node/dist/packages/contracts-node/typechain/core/ethers";
+import { MessageReceivedFromChildEvent as MessageReceivedFromChildArbitrum } from "@uma/contracts-node/dist/packages/contracts-node/typechain/core/ethers/ArbitrumParentMessenger";
+import { MessageSentToChildEvent as MessageSentToChildArbitrum } from "@uma/contracts-node/dist/packages/contracts-node/typechain/core/ethers/OptimismParentMessenger";
+import { RequestResolvedEvent } from "@uma/contracts-node/dist/packages/contracts-node/typechain/core/ethers/VotingV2";
+import { utils } from "ethers";
+import hre from "hardhat";
+import { logPricePublished } from "./BotLogger";
+import {
+  ARBITRUM_CHAIN_ID,
+  BLOCKS_WEEK_MAINNET,
+  Logger,
+  MonitoringParams,
+  getContractInstanceWithProvider,
+} from "./common";
+const { defaultAbiCoder } = utils;
+const ethers = hre.ethers;
+
+export async function publishPrices(logger: typeof Logger, params: MonitoringParams): Promise<void> {
+  const votingV2 = await getContractInstanceWithProvider<VotingV2Ethers>("VotingV2", params.provider);
+
+  const oracleHub = await getContractInstanceWithProvider<OracleHubEthers>("OracleHub", params.provider);
+
+  const arbitrumParentMessenger = await getContractInstanceWithProvider<ArbitrumParentMessenger>(
+    "Arbitrum_ParentMessenger",
+    params.provider
+  );
+
+  const currentBlockNumber = await params.provider.getBlockNumber();
+
+  const loopBack = params.blockLookbackResolution || BLOCKS_WEEK_MAINNET;
+  const searchConfig = {
+    fromBlock: currentBlockNumber - loopBack < 0 ? 0 : currentBlockNumber - loopBack,
+    toBlock: currentBlockNumber,
+    maxBlockLookBack: params.maxBlockLookBack,
+  };
+
+  // Find resolved events
+  const resolvedEvents = await paginatedEventQuery<RequestResolvedEvent>(
+    votingV2,
+    votingV2.filters.RequestResolved(null, null, null, null, null),
+    searchConfig
+  );
+
+  for (const event of resolvedEvents) {
+    const { identifier, time, ancillaryData, price } = event.args;
+
+    const encodedData = defaultAbiCoder.encode(["bytes32", "uint256", "bytes"], [identifier, time, ancillaryData]);
+
+    // Search for requested price from Arbitrum via ArbitrumParentMessenger
```

Comment:
Does this mean that we have to wait for the message to cross the slow bridge from arbitrum?

If so, I have two questions:
1. What about the speed-up bot that publishes the request on mainnet faster so it can be published back sooner? Does this currently handle those? Do we plan to?
2. How does this slow bridge message get finalized on mainnet? Is that the responsibility of a different bot?

Reply:
> 1. What about the speed-up bot that publishes the request on mainnet faster so it can be published back sooner? Does this currently handle those? Do we plan to?

No, this particular bot doesn't handle those (for now) because it's not possible to know the origin chain of a speedup request. To speedup a price request we use [requestPrice](https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/cross-chain-oracle/OracleHub.sol#L134) (even for Polygon that has it's own `OracleBaseTunnel` called `OracleRootTunnel`) and that function doesn't inform of the origin chain.

I will start with the speedup bot next week, I think that bot should store off-chain the information of the speeded up price requests and origin chains so the publisher bot can use them. The Speedup bot could:
- Speed up a price request
- Store the price request info + origin chain id in a db/cloud storage

Then I should edit the publisher bots to read the stored speeded up request and include them in the publishing logic

> 2. How does this slow bridge message get finalized on mainnet? Is that the responsibility of a different bot?

Sorry, not sure if I understand the question. The publisher bot's responsibility is calling `publishPrice` in the OracleHub or OracleRootTunnel when needed the the native bridges take care of the rest. Do you think there is something missing?